### PR TITLE
Preserve grok pattern ordering and add sort option

### LINKED ISSUE #40819: Ingest: Sort grok patterns when retrieving via REST API
**Describe the feature**: TIL about the [Retrieve patterns from Grok REST API](https://www.elastic.co/guide/en/elasticsearch/reference/6.6/grok-processor.html#grok-processor-rest-get), which is super awesome. However the output is fully random and hard to parse/check for humans.

I am not yet a hundred percent sure if this makes sense, but maybe ordering it initially when loading the patterns in the `Grok` class would result in a sorted map, that would be easier to read or go through - unless this API is only intended for machine consumption.

### PR DESCRIPTION
**Pull request by danhermann**
The [grok processor REST endpoint](https://www.elastic.co/guide/en/elasticsearch/reference/current/grok-processor.html#grok-processor-rest-get) returns the bundled grok patterns in an undetermined order. This PR changes the default behavior to return them in the order in which they were read from disk which preserves the natural grouping of related patterns such as "S3_ACCESS_LOG" and "ELB_URI" for AWS-related patterns though they share no common prefix. An option to return all patterns sorted by key was also added.

Resolves #40819.

### CODE REVIEW

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/default-distro

**Comment by danhermann** [comment]
@elasticmachine update branch

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/packaging-sample-windows

**Review by probakowski — COMMENTED** [review]
Thanks @danhermann for adding this, I left one serialization question and small possible improvement

File: modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/GrokProcessorGetAction.java [hunk 1]
```diff
@@ -108,11 +124,15 @@ public TransportAction(TransportService transportService, ActionFilters actionFi
...
             try {
-                listener.onResponse(new Response(GROK_PATTERNS));
+                listener.onResponse(new Response(getGrokPatternsResponse(GROK_PATTERNS, request.sorted())));
             } catch (Exception e) {
                 listener.onFailure(e);
             }
         }
+
+        static Map<String, String> getGrokPatternsResponse(Map<String, String> patterns, boolean sorted) {
+            return sorted ? new TreeMap<>(patterns) : patterns;
```
[the remark this answers is not in the document]
Comment (danhermann) [reply modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/GrokProcessorGetAction.java]: I shuffled around a few things in https://github.com/elastic/elasticsearch/pull/61671/commits/9d3a31a02b16699572e9ebd5a8bde2ed192f6607 to sort the patterns only once while retaining the ability to test the sort option. If you think it's useful, the sorted patterns member could be pulled into a singleton enum or the like for lazy loading. I was on the fence as to whether that would be particularly beneficial.

**Comment by danhermann** [comment]
@elasticmachine run elasticsearch-ci/bwc

**Comment by danhermann** [comment]
@elasticmachine update branch

**Review by probakowski — APPROVED** [review]
LGTM, thanks @danhermann!
